src/devices/machine/ncr53c90.cpp

```diff
@@ -1329,3 +1359,36 @@ void ncr53c94_device::check_drq()
 	else
 		ncr53c90_device::check_drq();
 }
+
+void ncr53c94_device::decrement_tcounter(int count)
+{
+	if (!dma_command)
+		return;
+
+	const uint32_t tcounter_mask = (config2 & S2FE) ? 0xffffff : 0xffff;
+	if (!(status & S_TC0))
+		tcounter = (tcounter - count) & tcounter_mask;
+	else
+		tcounter = 0;
+
+	if (tcounter == 0)
+		status |= S_TC0;
+
+	check_drq();
+}
+
+uint8_t ncr53c94_device::tcounter_hi2_r()
+{
+	// tcounter is 24-bit when the features bit is set, otherwise it returns the ID
+	if ((config2 & S2FE) == 0)
+		return (1 << 7) | (family_id << 3) | revision_level;
+
+	LOG("tcounter_hi2_r %02x (%s)\n", (tcounter >> 16) & 0xff, machine().describe_context());
+	return tcounter >> 16;
+}
+
+void ncr53c94_device::tcount_hi2_w(uint8_t data)
+{
+	tcount = (tcount & ~0xff0000) | (data << 16);
+	LOG("tcount_hi2_w %02x (%s)\n", data, machine().describe_context());
```

Comment:
I get kind of nervous with these binary inversions on implicitly typed integer literals, and shifting a value by more than the width of its type relying on C automatic promotion.  The behaviour is ABI-dependent.  Can you please cast the literals to the width you want before inverting them (e.g. `tcount & ~uint32_t(0xff00)`) and explicitly promote small integers before doing large shifts (e.g. `(uint32_t(data) << 16)`?

Reply:
Sure.

Reply:
Which ABI doesn't it work on?
I only know it can be problematic with u64.

Reply:
`int` is only guaranteed to be at least sixteen bits wide - this code will definitely behave incorrectly on LP32 targets. It's generally asking for trouble to rely on automatic promotion and implicit truncation. I've seen it produce hard-to-diagnose bugs and waste a lot of time.

Reply:
You're saying this won't work on some MAME targets?: for (int i=0;i<1000000;i++)
And what about shifting an 8 bit variable to the left?

Reply:
I am saying that it is not portable C, and MAME code is not to be written in a way that makes it difficult to get working on new targets.

Even though shifting a `uint8_t` to the left by eight positions will work, it’s just better to make a habit of not relying on automatic promotion of small types for arithmetic because of the trouble it *does* cause.  Automatic promotion for arithmetic in C is a misfeature.  It started as a leftover from BCPL, and now it’s something we’re stuck with.  It would have been better if the behaviour never existed in the first place.

Reply:
I doubt MAME will work at all if "int" is 16-bit, or on compilers that don't support automatic promotion.
IIRC we had a validity check to see that int is at least 32 bit, but I can't find it.